IPVGO: Support scripts playing against each other as each color on "No AI" boards

ficocelliguy opened the pull request:
Branch testable here: https://ficocelliguy.github.io/bitburner-src/

A few people in discord have asked for the ability to pit multiple IPvGO scripts against each other. This is the most recent:
https://discord.com/channels/415207508303544321/1221920484111814828/1330501717674168423

This PR adds support for making moves as the white player via the API. This can only be done on the "No AI" practice board, which gives no stats. To do this, additional promise handling was added so that the two scripts could listen for when the other script (or player via UI) has made a move.


In addition, documentation improvements and a new Go.analytics.getHistory() method were added to resolve this feedback:
Closes #1576

ficocelliguy committed:
IPVGO: Support scripts playing against each other as each color on "No AI" boards

ficocelliguy committed:
IPVGO: Update documentation

ficocelliguy committed:
IPvGO: remove extra node power addition

Sphyxis commented:
I look forward to implementing this in my script and so do many other people.

ficocelliguy committed:
IPvGO: Add protection against passing as white when not playing on a "No AI" board

ficocelliguy committed:
IPvGO: Add support for cheat API as white on "No AI" board

ficocelliguy committed:
IPvGO: Update documentation

ficocelliguy committed:
IPvGO: Unit test fixes

ficocelliguy committed:
IPvGO: Fix cheat success chance api type checking

ficocelliguy committed:
IPvGO: Fix error handling and promise updating when passing

ficocelliguy committed:
IPvGO: Fix chains not getting reset

ficocelliguy committed:
IPvGO: Set up a waiting promise for white when new games are started

ficocelliguy committed:
IPvGO: Fix promises getting overwritten

ficocelliguy committed:
IPvGO: Fix initial promise state; improve logging

ficocelliguy committed:
IPvGO: Fix promise handling and manual play

ficocelliguy committed:
IPvGO: Fix promise handling and manual play

ficocelliguy committed:
IPvGO: Fix promise handling and manual play

ficocelliguy committed:
IPvGO: Fix promise handling and manual play

ficocelliguy committed:
IPvGO: Fix manual passing

ficocelliguy committed:
lint

ficocelliguy committed:
IPvGO: Make scoring max controlled empty space consistent with visual controlled territory max space

ficocelliguy committed:
IPvGO: linting

d0sboots commented on src/ScriptEditor/NetscriptDefinitions.d.ts:
```diff
@@ -4333,11 +4333,13 @@ export interface GoAnalysis {
    * (This is intentionally expensive; you can derive this info from just getBoardState() )
    */
-  getValidMoves(boardState?: string[], priorBoardState?: string[]): boolean[][];
+  getValidMoves(boardState?: string[], priorBoardState?: string[], playAsWhite): boolean[][];
```
```suggestion
  getValidMoves(boardState?: string[], priorBoardState?: string[], playAsWhite?: boolean): boolean[][];
```

d0sboots commented on src/ScriptEditor/NetscriptDefinitions.d.ts:
```diff
@@ -4556,6 +4583,7 @@ export interface GoCheat {
   destroyNode(
     x: number,
     y: number,
+    playAsWhite?: boolean,
```
Optional: It probably makes more sense to do this here and wherever you can.
```suggestion
    playAsWhite = false,
```

d0sboots commented on src/Go/effects/netscriptGoImplementation.ts:
```diff
@@ -104,42 +120,53 @@ export function validateTurn(error: (s: string) => void, moveString = "") {
  */
-export async function getOpponentNextMove(logOpponentMove = true, logger: (s: string) => void) {
+export async function getOpponentNextMove(logOpponentMove = true, logger: (s: string) => void, playAsWhite = false) {
+  const nextMovePromise = playAsWhite ? "nextTurnForWhite" : "nextTurn";
```
This really solidifies my feeling that these should not be separate variables. Rather, there should be a single `nextTurn` object that is indexed by GoColor, or possibly a two-length array. The same probably makes sense for most of the things you've split out as separate variables.

ficocelliguy replied:
I reworked this to be a GoColor object, which is definitely cleaner

d0sboots commented on src/Go/SaveLoad.ts:
```diff
@@ -88,14 +90,18 @@ export function loadGo(data: unknown): boolean {
         ? { type: GoPlayType.move, x: previousMove[0], y: previousMove[1] }
         : { type: GoPlayType.pass, x: null, y: null },
     );
+    Go.nextTurnForWhite = cloneDeep(Go.nextTurn);
```
This is deeply suspicious to me. What happens if the ai is *not* GoOpponent.none?

Also, why even clone this? What's the issue with them referencing the same promise, if this usage was OK in the first place?

ficocelliguy replied:
This was part of testing for a bug where the wrong promise was being resolved, but I don't need it anymore. Good catch

d0sboots reviewed (commented):
I still need to take a deeper look at the nextTurn stuff, which is the part with the most magic (always has been).

ficocelliguy committed:
IPvGO: Consolidate nextTurn promises

d0sboots commented on src/Go/boardAnalysis/goAI.ts:
```diff
@@ -24,26 +25,36 @@ import { Go, GoEvents } from "../Go";
   // If AI is already taking their turn, return the existing turn.
   if (isAiThinking) {
-    return Go.nextTurn;
+    if (scriptPlayingAsWhite && currentTurnResolverAsWhite) {
```
The extra condition here worries me. The natural extension is `if (scriptPlayingAsWhite)`; it is trivially obvious that one of the two branches will be taken in that case. As it stands now, I can't be sure that the logic won't fall out of this if-block.

d0sboots commented on src/Go/boardAnalysis/goAI.ts:
```diff
@@ -24,26 +25,36 @@ import { Go, GoEvents } from "../Go";
   else {
     const currentMoveCount = Go.currentGame.previousBoards.length;
-    Go.nextTurn = getMove(boardState, GoColor.white, Go.currentGame.ai, useOfflineCycles).then(
+    Go.nextTurn[GoColor.black] = getMove(boardState, GoColor.white, Go.currentGame.ai, useOfflineCycles).then(
```
Not for the current PR, but: I feel like this (and other places) have a lot of color hardcoding that actually makes the code more complicated. With a little bit of refactoring, not only could the AI play as black (potentially), but also everything would probably be a lot simpler.

d0sboots commented on src/Go/boardAnalysis/goAI.ts:
```diff
@@ -83,35 +94,91 @@ export function makeAIMove(boardState: BoardState, useOfflineCycles = true): Pro
+    } else {
+      currentTurnResolver = handler;
+    }
+    // In cse the promise has become orphaned or overwritten, resolve if the other player makes a move
```
```suggestion
    // In case the promise has become orphaned or overwritten, resolve if the other player makes a move
```

d0sboots replied:
This is a maximum-level code smell. If orphaned/overwritten promises are occurring, that is *not* something we should be papering over with a mechanism like this, because it *will* manifest as weird unreproducible edge cases in people's scripts and this "fix" will only make it harder to track down the issues.

d0sboots commented on src/Go/boardAnalysis/goAI.ts:
```diff
@@ -83,35 +94,91 @@ export function makeAIMove(boardState: BoardState, useOfflineCycles = true): Pro
  * Resolves the current turn.
  * This is used for players manually playing against their script on the no-ai board.
  */
 export function resolveCurrentTurn() {
```
These don't need to be exported anymore (and thus, it's best that they aren't.)

d0sboots commented on src/Go/boardAnalysis/goAI.ts:
```diff
@@ -83,35 +94,91 @@ export function makeAIMove(boardState: BoardState, useOfflineCycles = true): Pro
+      : { type: GoPlayType.pass, x: null, y: null },
+  );
+
+  if (Go.currentGame.previousPlayer !== GoColor.black) {
```
```suggestion
  if (Go.currentGame.previousPlayer === GoColor.white) {
```
IMO it's easier to understand these in positive logic. Otherwise, I wonder what happens if two of them match.

d0sboots commented on src/Go/effects/netscriptGoImplementation.ts:
```diff
@@ -104,42 +120,53 @@ export function validateTurn(error: (s: string) => void, moveString = "") {
   // Only asynchronously log the opponent move if not disabled by the player
   if (logOpponentMove) {
-    return Go.nextTurn.then((move) => {
+    logger("Waiting for opponent to make their move...");
```
I thought this was handled elsewhere? I seem to remember a big discussion about this.

d0sboots replied:
I removed this in my changes, but feel free to make a case for putting it back if you think it should be there. (The part that seemed odd to be was that this says "waiting" but it won't log until *after* the opponent moves.)

d0sboots commented on src/Go/effects/netscriptGoImplementation.ts:
```diff
@@ -151,18 +178,25 @@ export async function getOpponentNextMove(logOpponentMove = true, logger: (s: st
   const boardState = _boardState || Go.currentGame;
+  const color = playAsWhite ? GoColor.white : GoColor.black;
+
+  // If the game is over, or if it is not your turn, there are no valid moves
```
This shouldn't be needed. (The reduce after this should do the same, and if it doesn't, then it ough to throw an error instead.)

d0sboots commented on src/Go/boardAnalysis/goAI.ts:
```diff
@@ -24,26 +25,36 @@ import { Go, GoEvents } from "../Go";
  * Retrieves a move from the current faction in response to the player's move
  */
-export function makeAIMove(boardState: BoardState, useOfflineCycles = true): Promise<Play> {
+export function makeAIMove(
```
Optional-ish: This method is increasingly misnamed. In cases like noAI there's no AI moves at all, and it's being called in response to *both* sides' moves. Something like `handleNextTurn` might be more appropriate.

d0sboots commented on src/Go/boardAnalysis/goAI.ts:
```diff
@@ -24,26 +25,36 @@ import { Go, GoEvents } from "../Go";
+let currentTurnResolverAsWhite: (() => void) | null = null;
 
 /**
  * Retrieves a move from the current faction in response to the player's move
```
This comment should be updated to describe the full responsibilities of this function. (It responds to a player's move by possibly using the AI to get a new move, and returns an appropriate promise based on the type of game.)

d0sboots commented on src/Go/ui/GoGameboardWrapper.tsx:
```diff
@@ -119,9 +117,9 @@ export function GoGameboardWrapper({ showInstructions }: GoGameboardWrapperProps
 
   async function takeAiTurn(boardState: BoardState) {
-    // If white is being played manually, halt and notify any scripts playing as black if present, instead of making an AI move
+    // Instead of making an AI move on the "No AI" opponent, notify any scripts waiting for their turn and then halt
```
This code shouldn't be needed; the script-play version doesn't have it, but relies on `makeAIMove` to do the right thing. It's better if they both go through the same codepath.

d0sboots reviewed (commented):
Every time I have to review this, I have to pull the full state of how the promise stuff works back into cache. XD I think I almost have a coherent set of suggestions to make, but I'm still ruminating. Pulling everything together into a single `update` method is the right plan.

d0sboots committed:
Refactoring and bugfixes

Fixes the issue where handicap stones are added on game load.
Better typing for error callbacks.
Throw errors instead of deadlocking on bad cheat usage.

d0sboots committed:
Massive redo of promise handling

This greatly simplifies the code by running everything through
handleNextTurn() (formerly makeAIMove()). The nextTurn promises for both
sides are handled identically, without special-casing. Even the built-in
AI is largely color-agnostic, with only one place that is hardcoded to
white.

d0sboots committed:
prettier/lint

d0sboots committed:
Return always-resolved gameOver promise after game end

This tweaked something unstable in Gameboard, the new condition is much
more stable for determining when to show the score dialog.

ficocelliguy committed:
IPvGO: Fix pass bug, add resetStats(), added one prior move to save

d0sboots commented:
I went through and resolved all the comments I handled in my PR, which was all of them XD

d0sboots committed:
Fix passing by reusing validateMove

d0sboots committed:
Merge branch 'ipvgo2' into IPvGO_support_scripts_playing_as_white_against_other_scripts

ficocelliguy committed:
Merge pull request #2 from d0sboots/IPvGO_support_scripts_playing_as_white_against_other_scripts

IPVGO: PR on a PR

ficocelliguy committed:
IPvGO: Merge

ficocelliguy committed:
Merge remote-tracking branch 'upstream/dev' into IPvGO_support_scripts_playing_as_white_against_other_scripts

ficocelliguy committed:
IPvGO: Docs

ficocelliguy committed:
IPvGO: Unit test fixes 1

ficocelliguy committed:
IPvGO: Unit test fixes 2

d0sboots committed:
Test fixes

ficocelliguy committed:
Merge pull request #3 from d0sboots/IPvGO_support_scripts_playing_as_white_against_other_scripts

Test fixes
